File: DataStructures/Tree/BST/python/BST.py
import sys

class Node:
	def __init__(self, data):
		self.data = data
		self.left = None
		self.right = None

class BST:
	def __init__(self):
		self.root = None

	def insert(self, root, data):
		if(root is None):
			return Node(data)
		if(data < root.data):
			root.left = self.insert(root.left, data)
		elif(data > root.data):
			root.right = self.insert(root.right, data)
		return root

	def preorder(self, root):
		if(root):
			print(root.data,end = " ")
			self.preorder(root.left)
			self.preorder(root.right)
		
	def inorder(self, root):
		if(root):
			print(root.data,end = " ")
			self.inorder(root.left)
			self.inorder(root.right)

	def postorder(self, root):
		if(root):
			print(root.data,end = " ")
			self.postorder(root.left)
			self.postorder(root.right)

	def search(self, root, data):
		if(root is None):
			return False
		if(root.data == data):
			return True
		elif(root.data < data):
			return self.search(root.right, data)
		else:
			return self.search(root.left, data)
		return False

	def getInorderSuccessor(self,temp):
		while(temp and temp.left):
			temp = temp.left;
		return temp;
	
	def delete(self, root, data):
		if(root is None):
			return root
		if(data < root.data):
			root.left = self.delete(root.left, data)
		elif(data > root.data):
			root.right = self.delete(root.right, data)
		else:
			#Node with one child or no child 
			if(root.left is None):
				temp = root.right
				root = None
				return temp
			elif(root.right is None):
				temp = root.left
				root = None
				return temp
			#Node with two child
			temp = self.getInorderSuccessor(root.right)
			root.data = temp.data
			root.right = self.delete(root.right, temp.data)

		return root


if __name__ == "__main__":

	sys.stdin = open('../input.txt', 'r')
	sys.stdout = open('../output.txt', 'w')

	n = sys.stdin.readline()
	a = sys.stdin.readline().split(" ") # [:-1] to eliminate reading of newline
	a = [int(i) for i in a]

	bst = BST();

	for i in a:
		bst.root = bst.insert(bst.root, i)

	for line in sys.stdin:

		if line[-1] == '\n':
			opt = line[:-1].split(" ")
		else:
			opt = line.split(" ")
		
		if opt[0] == "search":
			key = opt[1]
			isFound = bst.search(bst.root,int(key))
			if(isFound is True):
				print(key," is found.", end = "\n");
			else:
				print(key," is not found.", end = "\n");
		elif opt[0] == "delete":
			key = opt[1]
			bst.delete(bst.root, int(key))
		elif opt[0] == "inorder":
			print("Inorder traversal: ")
			bst.inorder(bst.root)
			print(end = "\n")
		elif opt[0] == "preorder":
			print("Preorder traversal: ")
			bst.preorder(bst.root)
			print(end = "\n")
		elif opt[0] == "postorder":
			print("Postorder traversal: ")
			bst.postorder(bst.root)
			print(end = "\n")

File: DataStructures/Tree/BST/cpp/BST.cpp
#include <stdlib.h>
#include <iostream>
#include <iomanip>
#include <string>
#include <algorithm>
#include <iterator>
#include <vector>

#define fast_io ios_base::sync_with_stdio(false);cin.tie(0);cout.tie(0);
#define endl "\n"

using namespace std;

struct Node{
	int data;
	struct Node *left, *right;
};

struct Node *newnode(int key){
	struct Node *node = new Node;
	node -> data = key;
	node -> left = node -> right = NULL;
	return node;
}

struct Node *insert(struct Node *root, int data){
	if(root == NULL)
		return newnode(data);
	if(data < root->data)
		root -> left = insert(root->left, data);
	else if(data > root->data)
		root -> right = insert(root->right, data);
	return root;
}

bool search(struct Node *root, int data){
	if(root == NULL)
		return false;
	if(root->data == data)
		return true;
	else if(root->data < data)
		return search(root->right, data);
	else
		return search(root->left, data);
	return false;
}

struct Node *getInorderSuccessor(struct Node *temp){
	while(temp && temp->left != NULL)
		temp = temp -> left;
	return temp;
}
struct Node *delete_(struct Node *root, int data){
	if(root == NULL)
		return root;
	if(data < root->data )
		root -> left = delete_(root->left, data);
	else if(data > root->data)
		root -> right = delete_(root->right, data);
	else{
		//Node with one child or no child 
		if(root -> left == NULL){
			struct Node *temp = root -> right;
			free(root);
			return temp;
		}
		else if(root -> right == NULL){
			struct Node *temp = root -> left;
			free(root);
			return temp;
		}
		//Node with two child
		struct Node *temp = getInorderSuccessor(root->right);
		root -> data = temp -> data;
		root -> right = delete_(root->right, temp -> data);
	}
	return root;
}

void preorder(struct Node* root){
	if(root){
		cout<<root -> data<<" ";
		preorder(root -> left);
		preorder(root -> right);
	}
}
void inorder(struct Node* root){
	if(root){
		inorder(root -> left);
		cout<<root -> data<<" ";
		inorder(root -> right);
	}
}
void postorder(struct Node* root){
	if(root){
		postorder(root -> left);
		postorder(root -> right);
		cout<<root -> data<<" ";
	}
}

int main(){
	fast_io;

	FILE *fp = freopen("../input.txt", "r", stdin);
	freopen("../output.txt", "w", stdout);

	int n, a[n];
	cin>>n;
	for(int i=0; i<n; i++)
		cin>>a[i];

	struct Node *root = NULL;

	for(int i=0; i<n; i++){
		root = insert(root, a[i]);
	}

	int t = 100000; // max input size
	while(t--){
		string opt;
		cin>>opt;
		if(opt == ""){
			break;  // eof
		}

		if(opt == "search"){
			int key;
			cin>>key;

			bool isFound = search(root,key); 
			if(isFound)
				cout<<key<<" is found."<<endl;
			else
				cout<<key<<" is not found"<<endl;
		}
		else if(opt == "delete"){
			int key;
			cin>>key;
			
			root = delete_(root, key);
		}
		else if(opt == "inorder"){
			cout<<"Inorder Traversal: "<<endl;
			inorder(root);
			cout<<endl;
		}
		else if(opt == "preorder"){
			cout<<"Preorder Traversal: "<<endl;
			preorder(root);
			cout<<endl;
		}
		else if(opt == "postorder"){
			cout<<"Postorder Traversal: "<<endl;
			inorder(root);
			cout<<endl;
		}
	}
	fclose(fp);
}